From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: fgei <fgei@gmail.com>
Date: Sun, 26 Feb 2023 10:39:38 +0100
Subject: [PATCH] Support for extended version name and increments

Adapted from GrapheneOS Vanadium patch 0004-Support-for-extended-version-name-and-increments.patch for the Vandium browser port.

---
 build/util/android_chrome_version.py |  8 ++++----
 build/util/version.py                |  8 +++++++-
 chrome/version.gni                   | 27 +++++++++++++++++++++++++++
 3 files changed, 38 insertions(+), 5 deletions(-)

diff --git a/build/util/version.py b/build/util/version.py
index c69969b91736c..922a4a2e6ca6c 100755
--- a/build/util/version.py
+++ b/build/util/version.py
@@ -136,6 +136,10 @@ def BuildParser():
                       help='Write substituted strings to FILE.')
   parser.add_argument('-t', '--template', default=None,
                       help='Use TEMPLATE as the strings to substitute.')
+  parser.add_argument(
+      '--increment',
+      default='0',
+      help='Version increment to append at version_name, add at version_code.')
   parser.add_argument('-x',
                       '--executable',
                       default=False,
@@ -209,9 +213,11 @@ def GenerateValues(options, evals):
   for key, val in evals.items():
     values[key] = str(eval(val, globals(), values))
 
+  values['INCREMENT'] = options.increment
   if options.os == 'android':
     android_chrome_version_codes = android_chrome_version.GenerateVersionCodes(
-        int(values['BUILD']), int(values['PATCH']), options.arch)
+        int(values['BUILD']), int(values['PATCH']), int(values['INCREMENT']),
+        options.arch)
     values.update(android_chrome_version_codes)
 
   return values
diff --git a/chrome/version.gni b/chrome/version.gni
index 7e7c4a838f656..2e335effd16fc 100644
--- a/chrome/version.gni
+++ b/chrome/version.gni
@@ -41,6 +41,33 @@ if (target_os == "android") {
   import("//build/config/android/config.gni")
   import("//build/config/chrome_build.gni")
 
+  declare_args() {
+    # Allows usage for extended versionName increment for android.
+    ext_version_enabled = false
+    # Increment positive integer to append at versionName,
+    # as well as to add on patch section of versionCode.
+    # String is used for utility at build/util/version.py.
+    ext_version_increment = ""
+  }
+
+  # Disallow setting extended version increment when disabled.
+  assert(ext_version_enabled || ext_version_increment == "")
+
+  if (ext_version_enabled) {
+    _version_dictionary_template = "full = \"@MAJOR@.@MINOR@.@BUILD@" +
+                                 ".@PATCH@.@INCREMENT@\" " +
+                                 "major = \"@MAJOR@\" minor = \"@MINOR@\" " +
+                                 "build = \"@BUILD@\" patch = \"@PATCH@\" " +
+                                 "increment = \"@INCREMENT@\" "
+
+    if (ext_version_increment != "") {
+      _script_arguments += [
+        "--increment",
+        ext_version_increment
+      ]
+    }
+  }
+
   _version_dictionary_template +=
       "chrome_version_code = " + "\"@CHROME_VERSION_CODE@\" " +
       "monochrome_version_code = \"@MONOCHROME_VERSION_CODE@\" " +
